Firestore/core/src/firebase/firestore/api/query_snapshot.mm

```diff
@@ -63,6 +67,91 @@ DocumentSnapshot snap(firestore_, document.key, document, from_cache,
   }
 }
 
+static DocumentChange::Type DocumentChangeTypeForChange(
+    const DocumentViewChange& change) {
+  switch (change.type()) {
+    case DocumentViewChange::Type::kAdded:
+      return DocumentChange::Type::Added;
+    case DocumentViewChange::Type::kModified:
+    case DocumentViewChange::Type::kMetadata:
+      return DocumentChange::Type::Modified;
+    case DocumentViewChange::Type::kRemoved:
+      return DocumentChange::Type::Removed;
+  }
+
+  HARD_FAIL("Unknown DocumentViewChange::Type: %s", change.type());
+}
+
+void QuerySnapshot::ForEachChange(
+    bool includeMetadataChanges,
+    const std::function<void(DocumentChange)>& callback) const {
+  if (includeMetadataChanges && snapshot_.excludes_metadata_changes()) {
+    ThrowInvalidArgument("To include metadata changes with your document "
+                         "changes, you must call "
+                         "addSnapshotListener(includeMetadataChanges:true).");
+  }
+
+  if (snapshot_.old_documents().empty()) {
+    // Special case the first snapshot because index calculation is easy and
+    // fast. Also all changes on the first snapshot are adds so there are also
+    // no metadata-only changes to filter out.
+    FSTDocument* lastDocument = nil;
+    size_t index = 0;
+    for (const DocumentViewChange& change : snapshot_.document_changes()) {
+      FSTDocument* doc = change.document();
+      SnapshotMetadata metadata = SnapshotMetadata(
+          /*hasPendingWrites=*/snapshot_.mutated_keys().contains(doc.key),
+          /*fromCache=*/snapshot_.from_cache());
+      DocumentSnapshot document =
+          DocumentSnapshot(firestore_, doc.key, doc, std::move(metadata));
```

Comment:
Nit: `SnapshotMetadata` consists of two primitives, moving it is no more efficient than copying (moving is only efficient for a resource with a handle, where copying a handle is significantly cheaper than copying the underlying resource).

Reply:
Ultimately, it's a matter of preference. I can see the argument that when logically ownership is transferred, `move` should always be applied, and if that's the approach you'd prefer to follow, that's fine with me.

My view is that, because `std::move` results in a no-op in this case, applying it may be misleading to the reader. My thought process would be something like `Hmm, I thought `SnapshotMetadata` is a simple POD, looks like that's not the case? Oh, looking at the class definition, it actually _is_ POD, so why is it being moved? Was that written deliberately, or did the author actually want to move something else?..` Basically, I'd avoid explicitly writing a no-op operation (in non-generic code), because I see it as redundant and potentially confusing.

Having said that, once again, it's more of a stylistic preference, and I don't feel very strongly about it.

Reply:
Gotcha'. I think your explanation makes sense and I'll make an attempt to follow it.  I do wish there was a straightforward rule to apply (e.g. without having to go look at SnapshotMetadata to determine its a POD), but until I come up with one I'll try to only std::move() when it's not a no-op.